crates/uv-requirements/src/pyproject.rs

```diff
@@ -1,133 +1,520 @@
+//! Reading from `pyproject.toml`
+//! * `project.{dependencies,optional-dependencies}`,
+//! * `tool.uv.sources` and
+//! * `tool.uv.workspace`
+//!
+//! and lowering them into a dependency specification.
+
+use std::collections::HashMap;
+use std::io;
+use std::ops::Deref;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+use glob::Pattern;
 use indexmap::IndexMap;
+use path_absolutize::Absolutize;
 use rustc_hash::FxHashSet;
 use serde::{Deserialize, Serialize};
-use std::str::FromStr;
+use thiserror::Error;
+use url::Url;
 
-use pep508_rs::Requirement;
-use pypi_types::LenientRequirement;
+use distribution_types::{ParsedUrlError, Requirement, RequirementSource, Requirements};
+use pep508_rs::{VerbatimUrl, VersionOrUrl};
+use uv_fs::Simplified;
+use uv_git::GitReference;
 use uv_normalize::{ExtraName, PackageName};
 
 use crate::ExtrasSpecification;
 
+#[derive(Debug, Error)]
+pub enum Pep621Error {
+    #[error(transparent)]
+    Pep508(#[from] pep508_rs::Pep508Error),
+    #[error("You need to specify a `[project]` section to use `[tool.uv.sources]`")]
+    MissingProjectSection,
+    #[error("pyproject.toml section is declared as dynamic, but must be static: `{0}`")]
+    CantBeDynamic(&'static str),
+    #[error("Failed to parse entry for: `{0}`")]
+    LoweringError(PackageName, #[source] UvSourcesLoweringError),
+}
+
+/// An error parsing and merging `tool.uv.sources` with
+/// `project.{dependencies,optional-dependencies}`.
+#[derive(Debug, Error)]
+pub enum UvSourcesLoweringError {
```

Comment:
Can we just call this `LoweringError`? I want to avoid having `Uv` in so many struct and abstraction names.